ydb/docs/en/core/reference/ydb-sdk/topic.md

```diff
@@ -776,6 +783,139 @@ All the metadata provided when writing a message is returned to a consumer with
 
   ```
 
+- Java
+
+  To take advantage of message metadata feature, build messages with builder. You can add `MetadataItem` to a message or set a `List` of `MetadataItem`s. Each item consists of a key of type `String` and a value of type `byte[]`:
+
+  ```java
+  writer.send(
+          Message.newBuilder()
+                  .addMetadataItem(new MetadataItem("meta-key", "meta-value".getBytes()))
+                  .addMetadataItem(new MetadataItem("another-key", "value".getBytes()))
+                  .build()
+  ```
+
+  While reading, metadata can be received from a `Message` with `getMetadataItems()` method:
+
+  ```java
+  Message message = reader.receive();
+  List<MetadataItem> metadata = message.getMetadataItems();
+  ```
+
+{% endlist %}
+
+### Write in transaction {#write-tx}
+
+{% list tabs %}
+
+- Java (sync)
+
+  [Example on GitHub](https://github.com/ydb-platform/ydb-java-examples/blob/develop/ydb-cookbook/src/main/java/tech/ydb/examples/topic/transactions/TransactionWriteSync.java)
+
+  Transaction can be set in `SendSettings` of `send` method while sending a message.
+  Such message will be written on transaction commit.
+
+  ```java
+  // creating session in table service
+  Result<Session> sessionResult = tableClient.createSession(Duration.ofSeconds(10)).join();
+  if (!sessionResult.isSuccess()) {
+      logger.error("Couldn't get session from pool: {}", sessionResult);
+      return; // retry or shutdown
+  }
+  Session session = sessionResult.getValue();
+  // creating transaction in table service
+  // this transaction is not yet active and has no id
+  TableTransaction transaction = session.createNewTransaction(TxMode.SERIALIZABLE_RW);
+
+  // do something else in transaction
+  transaction.executeDataQuery("SELECT 1").join();
+  // now transaction is active and has id
+  // analyzeQueryResultIfNeeded();
+
+  writer.send(
+          Message.newBuilder()
+                  .setData(messageString.getBytes())
+                  .build(),
+          SendSettings.newBuilder()
+                  .setTransaction(transaction)
+                  .build(),
+          timeoutSeconds,
+          TimeUnit.SECONDS
+  );
+
+  // flush to wait until all messages reach server
+  writer.flush();
+
+  Status commitStatus = transaction.commit().join();
+  analyzeCommitStatus(commitStatus);
+  ```
+
+  Transaction requirements:
+  It should be an active transaction (that has id) from one of YDB services. I.e. Table or Query.
```

Comment:
this could use links to more information about these services

Reply:
We don't have QueryService documentation yet and the is no documentation about Table Service overall, just particular calls

Reply:
@pnv1 at least to their grpc definitions then

Reply:
I don't see any use of it, actually. I'd better add links to TableTransaction and QueryTransaction in SDK if you don't mind. Makes more sense for me